src/Course/FileIO.hs

```diff
@@ -81,54 +81,52 @@ the contents of c
 
 -- Given the file name, and file contents, print them.
 -- Use @putStrLn@.
-printFile ::
-  FilePath
-  -> Chars
-  -> IO ()
-printFile =
-  error "todo: Course.FileIO#printFile"
+printFile :: FilePath -> Chars -> IO ()
+printFile path contents =
+    putStrLn $ "============ " ++ path ++ "\n" ++ contents
 
 -- Given a list of (file name and file contents), print each.
 -- Use @printFile@.
-printFiles ::
-  List (FilePath, Chars)
-  -> IO ()
-printFiles =
-  error "todo: Course.FileIO#printFiles"
+printFiles :: List (FilePath, Chars) -> IO ()
+printFiles files = execIoNoResult printTuple files
+  where
+    printTuple (path, chars) = printFile path chars
 
 -- Given a file name, return (file name and file contents).
 -- Use @readFile@.
-getFile ::
-  FilePath
-  -> IO (FilePath, Chars)
-getFile =
-  error "todo: Course.FileIO#getFile"
+getFile :: FilePath -> IO (FilePath, Chars)
+getFile path = ((,) path) <$> readFile path
```

Comment:
This is a bit tricky to read. It's okay, but maybe this is clearer?

```
getFile path = readFile path >>= \contents -> (path, contents)
```

Reply:
Definitely!

Reply:
```

src/Course/FileIO.hs:98:16-62: error:
    • Couldn't match type ‘List Char’ with ‘(FilePath, Chars)’
      Expected type: IO (FilePath, Chars)
        Actual type: IO Chars
    • In the expression:
        readFile path >>= \ contents -> (path, contents)
      In an equation for ‘getFile’:
          getFile path = readFile path >>= \ contents -> (path, contents)
   |
98 | getFile path = readFile path >>= \contents -> (path, contents)
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Course/FileIO.hs:98:47-62: error:
    • Couldn't match type ‘(,) FilePath’ with ‘IO’
      Expected type: IO Chars
        Actual type: (FilePath, Chars)
    • In the expression: (path, contents)
      In the second argument of ‘(>>=)’, namely
        ‘\ contents -> (path, contents)’
      In the expression: readFile path >>= \ contents -> (path, contents)
   |
98 | getFile path = readFile path >>= \contents -> (path, contents)
   |                                               ^^^^^^^^^^^^^^^^
```

:thinking:

I'll try to rewrite it clearer somehow.

Reply:
`getFile path = readFile path >>= \contents -> return (path, contents)` 

Does that look right?

Reply:
I thought it'd be better to use the most specific function available. i.e. if it's only monads, use return, if it's more general than that, use pure. Is that a good policy?

Reply:
It's usually best to do the opposite... ie use the most general function one can get away with. Learning is probably the only reason not to do this, or if you want to make something super clear.

If you have a `Monad`, then you also have an `Applicative` (coz it has to be in order to be `Monad`) and you have a `Functor`. That's the way the hierarchy goes for  this line of the tree. If your function only uses `<$>`, then you don't need `Monad` or `Applicative` as a constraint.

The reason for this is the more general the function is, the less ways there are to define that function.

Let's think about this function `someFunc :: a -> b -> a`... there's literally only one way to define that function. Or, the function `someOtherFunc :: a -> a` same thing — there's only one way you can define that that the typechecker will accept (if we ignore the bottom: `undefined` or `error`). The types guide us to the right function... however, if we make it more specific, say use `String`... someOtherFunctForString :: String -> String` there are literally hundreds of ways you can write this and the typechecker will accept it.

That's one reason. The other is we get reuse for free the more general we are.